server/src/main/java/org/opensearch/index/store/BaseRemoteSegmentStoreDirectory.java

```diff
@@ -0,0 +1,33 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.store;
+
+import org.apache.lucene.store.Directory;
+import org.apache.lucene.store.FilterDirectory;
+
+import java.io.IOException;
+import java.nio.file.Path;
+
+/**
+ * BaseRemoteSegmentsStoreDirectory - abstract class created to extend on the functionality of RemoteSegmentStoreDirectory
+ * Can add other public methods of RemoteSegmentStoreDirectory here as well such as init() and readLatestMetadata()
+ */
+
+public abstract class BaseRemoteSegmentStoreDirectory extends FilterDirectory {
```

Comment:
What about doing something like this instead?

```java
public class BlockedSegmentStoreDirectory extends RemoteSegmentStoreDirectory {

    TransferManager transferManager;
    FSDirectory localStoreDir;
   
   public BlockedSegmentStoreDirectory(RemoteDirectory remoteDataDirectory,
                                        RemoteDirectory remoteMetadataDirectory,
                                        RemoteStoreLockManager mdLockManager,
                                        ThreadPool threadPool,
                                        ShardId shardId,
                                        FSDirectory localStoreDir,
                                        FileCache fileCache) throws IOException {
        super(remoteDataDirectory, remoteMetadataDirectory, mdLockManager, threadPool, shardId);
        this.localStoreDir = localStoreDir;
        this.transferManager = new TransferManager(remoteDataDirectory.blobContainer, fileCache);
    }

    @Override
    public IndexInput openInput(String name, IOContext context) throws IOException {
        UploadedSegmentMetadata fileMetadata = getSegmentsUploadedToRemoteStore().get(name);
        return new OnDemandBlockSnapshotIndexInput(fileMetadata, localStoreDir, transferManager);
    }
```

yes we would need to break RemoteSegmentStoreDirectory in its current state by removing its final modifier to make this work but we conform to the Directory API as is.

Then we should be able to construct the necessary FileInfo here and reuse OnDemandBlockSnapshotIndexInput?

Reply:
@sachinpkale any implications of removing the final modifier from RemoteSegmentStoreDirectory ?

Reply:
Updated in the latest revisions